synapse/storage/databases/main/deviceinbox.py

```diff
@@ -737,31 +737,28 @@ async def _remove_hidden_devices_from_device_inbox(
         def _remove_hidden_devices_from_device_inbox_txn(
             txn: LoggingTransaction,
         ) -> int:
-            """stream_id is not unique
-            we need to use an inclusive `stream_id >= ?` clause,
-            since we might not have deleted all hidden device messages for the stream_id
-            returned from the previous query
+            """Retrieve the list of hidden devices that appear in the device_inbox table,
+            ordered by user_id so the query uses the index on the devices table.
 
             Then delete only rows matching the `(user_id, device_id, stream_id)` tuple,
             to avoid problems of deleting a large number of rows all at once
             due to a single device having lots of device messages.
             """
 
-            last_stream_id = progress.get("stream_id", 0)
+            last_user_id = progress.get("user_id", "")
```

Comment:
the problem here is that we might pick up a half-finished copy of the migration from the old implementation, and because we're dong things in a different order now, this won't work. We'll need to start from scratch again (which might be most easily achieved by using a differently-named field to store the progress).